src/emu/emuopts.cpp

```diff
@@ -879,111 +889,176 @@ emu_options::software_options emu_options::evaluate_initial_softlist_options(con
-						//
-						// example from SMS:
-						//
-						//  <software name = "alexbmx">
-						//      ...
-						//      <sharedfeat name = "ctrl1_default" value = "paddle" />
-						//  </software>
-						for (const software_info_item &fi : swinfo->shared_features())
-						{
-							const std::string default_suffix = "_default";
-							if (fi.name().size() > default_suffix.size()
-								&& fi.name().compare(fi.name().size() - default_suffix.size(), default_suffix.size(), default_suffix) == 0)
-							{
-								std::string slot_name = fi.name().substr(0, fi.name().size() - default_suffix.size());
-								results.slot[slot_name] = fi.value();
-							}
-						}
+						// if we couldn't find the requirement, bail
+						if (!requirement_swdev || requirement_swdevs_found != 1)
+							return { };
+
+						// push the resolved identifier onto the stack
+						software_identifier_stack.push(std::make_tuple(requirement_swdev, requirement_software_name));
 					}
 				}
-				if (compatible)
-					break;
 			}
 
-			if (!compatible)
+			// identify other shared features specified as '<<slot name>>_default'
+			//
+			// example from SMS:
+			//
+			//  <software name = "alexbmx">
+			//      ...
+			//      <sharedfeat name = "ctrl1_default" value = "paddle" />
+			//  </software>
+			for (const software_info_item &fi : swinfo->shared_features())
 			{
-				software_list_device::display_matches(config, nullptr, software_name);
-
-				// The text of this options_error_exception() is then passed to osd_printf_error() in cli_frontend::execute().  Therefore, it needs
-				// to be human readable text.  We want to snake through a message about software incompatibility while being silent if that is not
-				// the case.
-				//
-				// Arguably, anything related to user-visible text should really be done within src/frontend.  The invocation of
-				// software_list_device::display_matches() should really be done there as well
-				if (!found)
-					throw options_error_exception("");
-				else
-					throw options_error_exception("Software '%s' is incompatible with system '%s'\n", software_name, m_system->name);
+				const std::string default_suffix = "_default";
+				if (fi.name().size() > default_suffix.size()
+					&& fi.name().compare(fi.name().size() - default_suffix.size(), default_suffix.size(), default_suffix) == 0)
+				{
+					std::string slot_name = fi.name().substr(0, fi.name().size() - default_suffix.size());
+					result.slot[slot_name] = fi.value();
+				}
 			}
 		}
 	}
-	return results;
+
+	// only return something if we're compatible
+	return is_compatible
+		? result
+		: std::optional<emu_options::software_options>();
+}
+
+
+//-------------------------------------------------
+//  parse_software_identifier
+//-------------------------------------------------
+
+std::tuple<std::string_view, std::string_view> emu_options::parse_software_identifier(std::string_view software_identifier)
```

Comment:
I don’t think this needs to be a member function at all – it makes the header more fragile for no reason.